### ISSUE TITLE
`no-use-before-define` fails on destructuring assignment in 2.0.0-beta.3

### ISSUE THREAD

**Issue by mikestopcontinues** [eslint/eslint#5135]
After updating (to fix another issue), given this:

```js
const {main, side, ...props} = this.props;
```

I get this:

```shell
  72:12  warning  'main' was used before it was defined  no-use-before-define
  72:23  warning  'side' was used before it was defined    no-use-before-define
  72:35  warning  'props' was used before it was defined      no-use-before-define
```

**Comment by eslintbot** [comment]
@mikestopcontinues Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by gyandeeps** [comment]
Please answer all the question from the above comment.

**Comment by mikestopcontinues** [comment]
```shell
$ eslint -v
v2.0.0-beta.3
```

.eslintrc:

```json
{
  "parser": "babel-eslint",

  "rules": {
    "comma-dangle": [1, "always-multiline"],
    "no-cond-assign": [1, "except-parens"],
    "no-console": 0,
    "no-constant-condition": 1,
    "no-control-regex": 1,
    "no-debugger": 1,
    "no-dupe-args": 1,
    "no-dupe-keys": 1,
    "no-duplicate-case": 0,
    "no-empty-character-class": 1,
    "no-empty": 1,
    "no-ex-assign": 1,
    "no-extra-boolean-cast": 1,
    "no-extra-parens": 0,
    "no-extra-semi": 1,
    "no-func-assign": 1,
    "no-inner-declarations": [1, "functions"],
    "no-invalid-regexp": 1,
    "no-irregular-whitespace": 1,
    "no-negated-in-lhs": 1,
    "no-obj-calls": 1,
    "no-regex-spaces": 1,
    "no-reserved-keys": 0,
    "no-sparse-arrays": 1,
    "no-unexpected-multiline": 1,
    "no-unreachable": 1,
    "use-isnan": 1,
    "valid-jsdoc": 1,
    "valid-typeof": 1,

    "accessor-pairs": 0,
    "complexity": 0,
    "consistent-return": 1,
    "curly": [1, "all"],
    "default-case": 0,
    "dot-notation": [1, { "allowKeywords": true, "allowPattern": "" }],
    "dot-location": [1, "property"],
    "eqeqeq": 1,
    "guard-for-in": 0,
    "no-alert": 1,
    "no-caller": 1,
    "no-div-regex": 1,
    "no-else-return": 1,
    "no-empty-label": 1,
    "no-eq-null": 0,
    "no-eval": 1,
    "no-extend-native": 1,
    "no-extra-bind": 1,
    "no-fallthrough": 0,
    "no-floating-decimal": 1,
    "no-implied-eval": 1,
    "no-iterator": 1,
    "no-labels": 1,
    "no-lone-blocks": 1,
    "no-loop-func": 1,
    "no-multi-spaces": 1,
    "no-multi-str": 1,
    "no-native-reassign": 1,
    "no-new-func": 1,
    "no-new-wrappers": 1,
    "no-new": 1,
    "no-octal-escape": 1,
    "no-octal": 1,
    "no-param-reassign": 0,
    "no-process-env": 0,
    "no-proto": 1,
    "no-redeclare": 1,
    "no-return-assign": 1,
    "no-script-url": 1,
    "no-self-compare": 1,
    "no-sequences": 1,
    "no-throw-literal": 1,
    "no-unused-expressions": 0,
    "no-void": 0,
    "no-warning-comments": 0,
    "no-with": 1,
    "radix": 1,
    "vars-on-top": 1,
    "wrap-iife": [1, "inside"],
    "yoda": [1, "never"],

    "strict": [1, "never"],

    "no-catch-shadow": 0,
    "no-delete-var": 1,
    "no-label-var": 1,
    "no-shadow-restricted-names": 1,
    "no-shadow": 1,
    "no-undef-init": 1,
    "no-undef": 1,
    "no-undefined": 1,
    "no-unused-vars": [1, { "vars": "local", "args": "after-used" }],
    "no-use-before-define": 0,
    "no-empty-pattern": 1,

    "handle-callback-err": 1,
    "no-mixed-requires": 1,
    "no-new-require": 1,
    "no-path-concat": 1,
    "no-process-exit": 1,
    "no-restricted-modules": [1, ""],
    "no-sync": 1,

    "array-bracket-spacing": [1, "never"],
    "brace-style": [1, "1tbs", {"allowSingleLine": true}],
    "camelcase": [1, { "properties": "always" }],
    "comma-spacing": [1, { "before": false, "after": true }],
    "comma-style": [1, "last"],
    "computed-property-spacing": 0,
    "consistent-this": 0,
    "eol-last": 1,
    "func-names": 1,
    "func-style": 0,
    "indent": [1, 2],
    "key-spacing": [1, { "beforeColon": false, "afterColon": true }],
    "linebreak-style": 0,
    "max-nested-callbacks": [0, 3],
    "new-parens": 1,
    "newline-after-var": 0,
    "no-array-constructor": 1,
    "no-continue": 1,
    "no-inline-comments": 0,
    "no-lonely-if": 1,
    "no-mixed-spaces-and-tabs": 1,
    "no-multiple-empty-lines": [1, { "max": 1 }],
    "no-nested-ternary": 0,
    "no-new-object": 1,
    "no-spaced-func": 1,
    "no-ternary": 0,
    "no-trailing-spaces": 0,
    "no-underscore-dangle": 0,
    "no-unneeded-ternary": 1,
    "one-var": [1, "never"],
    "operator-assignment": [1, "always"],
    "padded-blocks": [0, "never"],
    "quote-props": [0, "as-needed"],
    "quotes": [1, "single"],
    "semi-spacing": [1, { "before": false, "after": true }],
    "semi": [1, "always"],
    "sort-vars": 0,
    "space-after-keywords": 0,
    "space-before-blocks": [1, "always"],
    "space-before-function-paren": [1, "never"],
    "space-in-parens": [1, "never"],
    "space-infix-ops": 1,
    "keyword-spacing": 1,
    "space-unary-ops": 0,
    "spaced-comment": [1, "always"],
    "wrap-regex": 1,

    "constructor-super": 1,
    "no-this-before-super": 1,
    "no-var": 1,
    "prefer-const": 1,

    "max-depth": [0, 3],
    "max-len": 0,
    "max-params": 0,
    "max-statements": 0,
    "no-bitwise": 1,
    "no-plusplus": 1,

    "object-shorthand": 0,
    "babel/object-shorthand": [1, "always"],

    "generator-star-spacing": 0,
    "babel/generator-star-spacing": [1, "after"],

    "new-cap": 0,
    "babel/new-cap": 1,

    "object-curly-spacing": 0,
    "babel/object-curly-spacing": [1, "never"],

    "jsx-quotes": [1, "prefer-double"],
    "react/display-name": [1, {
      "acceptTranspilerName": true
    }],
    "react/jsx-boolean-value": 1,
    "react/jsx-no-duplicate-props": 1,
    "react/jsx-no-undef": 1,
    "react/jsx-sort-prop-types": 1,
    "react/jsx-sort-props": 0,
    "react/jsx-uses-react": 1,
    "react/jsx-uses-vars": 1,
    "react/no-danger": 0,
    "react/no-did-mount-set-state": 1,
    "react/no-did-update-set-state": 1,
    "react/no-multi-comp": 1,
    "react/no-unknown-property": 1,
    "react/prop-types": 1,
    "react/react-in-jsx-scope": 1,
    "react/self-closing-comp": 1,
    "react/sort-comp": 1,
    "react/wrap-multilines": 0,
    "react/prefer-es6-class": 2,
  },

  "plugins": [
    "babel",
    "react",
  ],

  "settings": {
    "ecmascript": 7,
    "jsx": true
  },

  "env": {
    "es6": true,
    "browser": true,
    "node": true,
    "mocha": true
  }
}
```

**Comment by ilyavolodin** [comment]
Can you try it without `babel-eslint`?

**Comment by zandaqo** [comment]
`const {main, side, ...props} = this.props;`

As far as I know, the rest pattern for object destructuring is not in the current (ES2015/ES6) standard; chances are Espree simply doesn't support it.

**Comment by ilyavolodin** [comment]
Espree supports object rest/spread (that's the only post ES6 feature we support). You need to enable `experimentalObjectRestSpread` in the config `ecmaFeatures` property.

**Comment by mikestopcontinues** [comment]
After removing "babel-eslint", given this input:

```js
const {a, b, ...c} = window;
```

I get:

```shell
 14:21  error  Parsing error: Unexpected token ..
```

But given this input, everything's ok:

```js
const {a, b, c} = window;
```

**Comment by ilyavolodin** [comment]
@eslint/eslint-team Does Espree supposed to be able to parse this?

**Comment by gyandeeps** [comment]
Yes it works. This is how the config should be setup

```json
    "parserOptions": {
        "ecmaVersion": 6,
        "ecmaFeatures":{
            "experimentalObjectRestSpread": true
        }
    }
```

**Comment by mysticatea** [comment]
Yup.

**.eslintrc**

```yml
parserOptions:
    ecmaVersion: 6
    ecmaFeatures:
        experimentalObjectRestSpread: true
rules:
    no-use-before-define: 2
    no-unused-vars: 2
```

**test.js:**

```js
const {a, b, ...c} = window;
```

**Result:**

```
$ eslint test.js --no-ignore

C:\Users\t-nagashima.AD\Documents\GitHub\eslint\test.js
  1:8   error  'a' is defined but never used  no-unused-vars
  1:11  error  'b' is defined but never used  no-unused-vars
  1:17  error  'c' is defined but never used  no-unused-vars

✖ 3 problems (3 errors, 0 warnings)
```

This result is correct.

**Comment by ilyavolodin** [comment]
@mikestopcontinues looks like original issue is `babel-eslint` specific. You might want to open an issue there.

**Comment by mikestopcontinues** [comment]
Thanks!

**Comment by stoplion** [comment]
According [this](https://github.com/babel/babel-eslint#note-you-dont-need-to-use-babel-eslint-if-you-are-only-using-es6es2015-eslint-actually-supports-es6es7-jsx-and-object-restspread-by-default-now),  babel-eslint is no longer needed and yes this issue this happens for me. 

```
module.exports = {
  "env": {
    "browser": true,
    "commonjs": true,
    "es6": true
  },
  "extends": "eslint:recommended",
  "installedESLint": true,
  "parserOptions": {
    "ecmaVersion": 6,
    "ecmaFeatures": {
      "experimentalObjectRestSpread": true,
      "jsx": true
    },
    "sourceType": "module"
  },
  "plugins": [
    "react"
  ],
  "rules": {
    "indent": [
      "error",
      2,
      {"SwitchCase": 1}
    ],
    "linebreak-style": [
      "error",
      "unix"
    ],
    "quotes": [
      "error",
      "single"
    ],
    "semi": [
      "error",
      "always"
    ]
  }
```

in my code..
`let {moduleID} = action.data;`
`moduleID is defined but never used`

`eslint v3.2.2`

**Comment by IanVS** [comment]
@stoplion, That looks correct.  You've destructured `moduleID`, but not used it anywhere.  That does not seem related to this issue.
